Engine/Animation/AnimController.cpp

```diff
@@ -0,0 +1,198 @@
+#include "AnimController.h"
+#include "Main/Application.h"
+#include "Module/ModuleResourceManager.h"
+#include "Module/ModuleTime.h"
+#include "Helper/Utils.h"
+
+#include <math.h>
+#include <algorithm>
+
+void AnimController::Init()
+{
+	animation_time = (anim->frames/anim->frames_per_second )* 1000;
+	for (size_t i = 0; i < anim->keyframes[0].channels.size(); ++i)
+	{
+		auto & channel = anim->keyframes[0].channels[i];
+		auto it = std::find_if(sk->skeleton.begin(), sk->skeleton.end(), [channel](const Skeleton::Joint & joint)
+		{
+			return joint.name == channel.name;
+		});
+		if (it != sk->skeleton.end())
+		{
+			Skeleton::Joint joint = (*it);
+			while (joint.parent_index < sk->skeleton.size())
+			{
+				joint = sk->skeleton[joint.parent_index];
+				auto it_channel = std::find_if(anim->keyframes[0].channels.begin(), anim->keyframes[0].channels.end(), [&joint](const Animation::Channel & parent_channel)
+				{
+					return joint.name == parent_channel.name;
+				});
+				if (it_channel != anim->keyframes[0].channels.end())
+				{
+					channel_hierarchy_cache[i].push_back(it_channel - anim->keyframes[0].channels.begin());
+				}
+			}
+		}
+
+	}
+}
+
+
+void AnimController::Play()
+{ 
+	current_time = 0.f;
+	playing = true;
+}
+
+void AnimController::Stop()
+{
+	playing = false;
+}
+
+void AnimController::Update()
+{
+	if (!playing)
+	{
+		return;
+	}
+
+	current_time = current_time + App->time->delta_time;
+	if (current_time >= animation_time)
+	{
+		if (loop) 
+		{
+
+			current_time = current_time % animation_time;
+		}
+		else
+		{
+			playing = false;
+		}
+	}
+
+}
+
+bool AnimController::GetTranslation(const std::string& channel_name, float3 & pos)
+{
+	float current_sample = (current_time*(anim->frames - 1)) / animation_time;
+	int current_keyframe = math::FloorInt(current_sample);
+
+	int next_keyframe = (current_keyframe + 1) % (int)anim->frames;
+
+	if (current_keyframe >= anim->keyframes.size() || next_keyframe >= anim->keyframes.size())
+	{
+		return false;
+	}
+	float3 current_translation;
+	float3 next_translation;
+
+	bool channel_found = false;
+	size_t i = 0;
+	while (!channel_found && i < anim->keyframes[current_keyframe].channels.size())
+	{
+		if (anim->keyframes[current_keyframe].channels[i].name == channel_name)
+		{
+			channel_found = true;
+			if (!anim->keyframes[current_keyframe].channels[i].is_translated)
+			{
+				return false;
+			}
+			current_translation = anim->keyframes[current_keyframe].channels[i].translation;
+			next_translation = anim->keyframes[next_keyframe].channels[i].translation;
+		}
+		++i;
+	}
+
+	if (channel_found)
+	{
+		float delta = current_sample - current_keyframe;
+		pos = Utils::Interpolate(current_translation, next_translation, delta);
+	}
+
+	return channel_found;
+}
+
+bool AnimController::GetRotation(const std::string& channel_name, Quat & rot)
```

Comment:
Code style: Quat& and also change rot for rotation?(I know can be undestood already but.. code style and readability)